src/rulesets/Base/broadcast.jl

```diff
@@ -0,0 +1,370 @@
+using Base.Broadcast: Broadcast, broadcasted, Broadcasted
+const RCR = RuleConfig{>:HasReverseMode}
+
+function rrule(::typeof(copy), bc::Broadcasted)
+    uncopy(Δ) = (NoTangent(), Δ)
+    return copy(bc), uncopy
+end
+
+# Skip AD'ing through the axis computation
+function rrule(::typeof(Broadcast.instantiate), bc::Broadcasted)
+    uninstantiate(Δ) = (NoTangent(), Δ)
+    return Broadcast.instantiate(bc), uninstantiate
+end
+
+_print(args...) = printstyled("CR: ", join(args, " "), color=:magenta) # nothing # 
+
+#####
+##### Split broadcasting
+#####
+
+# For `z = g.(f.(xs))`, this finds `y = f.(x)` eagerly because the rules for either `f` or `g` may need it,
+# and we don't know whether re-computing `y` is cheap. 
+# (We could check `f` first like `sum(f, x)` does, but checking whether `g` needs `y` is tricky.)
+
+function rrule(cfg::RCR, ::typeof(broadcasted), f::F, args::Vararg{Any,N}) where {F,N}
+    T = Broadcast.combine_eltypes(f, args)
+    if T === Bool  # TODO use nondifftype here
+        # 1: Trivial case: non-differentiable output, e.g. `x .> 0`
+        _print("split_bc_trivial", f)
```

Comment:
just marking this as needing to be commented out before final version.
Or replaced with `@debug`

Reply:
Now `@debug`. Not entirely sure how safe that is, re 2nd derivatives etc, but it is convenient for now.

Reply:
should be safe on modern version of julia